python/plugins/processing/algs/qgis/Clip.py

```diff
@@ -60,76 +60,68 @@ def defineCharacteristics(self):
-                        QgsFeatureRequest().setFilterFid(i)).nextFeature(
-                            inFeatB)
-                    tmpGeom = QgsGeometry(inFeatB.geometry())
-                    if tmpGeom.intersects(geom):
-                        found = True
-                        if first:
-                            outFeat.setGeometry(QgsGeometry(tmpGeom))
-                            first = False
-                        else:
-                            cur_geom = QgsGeometry(outFeat.geometry())
-                            new_geom = QgsGeometry(cur_geom.combine(tmpGeom))
-                            if new_geom.isGeosEmpty() or not new_geom.isGeosValid():
-                                ProcessingLog.addToLog(ProcessingLog.LOG_ERROR,
-                                                       self.tr('GEOS geoprocessing error: One or '
-                                                               'more input features have invalid '
-                                                               'geometry.'))
-                                break
-
-                            outFeat.setGeometry(QgsGeometry(new_geom))
-                if found:
-                    cur_geom = QgsGeometry(outFeat.geometry())
-                    new_geom = QgsGeometry(geom.intersection(cur_geom))
-                    if new_geom.wkbType() == QGis.WKBUnknown or QgsWKBTypes.flatType(new_geom.geometry().wkbType()) == QgsWKBTypes.GeometryCollection:
-                        int_com = QgsGeometry(geom.combine(cur_geom))
-                        int_sym = QgsGeometry(geom.symDifference(cur_geom))
-                        new_geom = QgsGeometry(int_com.difference(int_sym))
-                        if new_geom.isGeosEmpty() or not new_geom.isGeosValid():
-                            ProcessingLog.addToLog(ProcessingLog.LOG_ERROR,
-                                                   self.tr('GEOS geoprocessing error: One or more '
-                                                           'input features have invalid geometry.'))
-                            continue
-                    try:
-                        outFeat.setGeometry(new_geom)
-                        outFeat.setAttributes(attrs)
-                        writer.addFeature(outFeat)
-                    except:
+        # first build up a list of clip geometries
+        clip_geoms = []
+        for maskFeat in maskLayer.getFeatures(QgsFeatureRequest().setSubsetOfAttributes([])):
+            clip_geoms.append(QgsGeometry(maskFeat.geometry()))
+
+        if len(clip_geoms) > 1:
+            clip_geom = QgsGeometry.unaryUnion(clip_geoms)
+        else:
+            clip_geom = clip_geoms[0]
+        # use prepared geometries for faster insection tests
+        engine = QgsGeometry.createGeometryEngine(clip_geom.geometry())
+        engine.prepareGeometry()
+
+        input_features = [f for f in sourceLayer.getFeatures(QgsFeatureRequest().setFilterRect(clip_geom.boundingBox()))]
```

Comment:
I understand... 
can be solved with multiple iterators (having a precise count), but the risk is at the end to do setFilterRect on multiple masks loosing the advantage to do it in a single step (at least for the fetching)

as I understand the main advantage here is to apply intersect with the union in the for loop, I can't evaluate the cost of multiple fetches instead of a single one potentially bigger than necessary.

btw it's a simple speculation, and the actual solution is wonderful for me :) I imagined only use cases like landscape analysis where the clip mask polygons are spatially sparse fetching most of features in the middle of these masks.